crates/ide-db/src/search.rs

```diff
@@ -126,8 +126,75 @@ impl SearchScope {
         SearchScope { entries: source_root.iter().map(|id| (id, None)).collect() }
     }
 
+    /// Build a search scope spanning the given module but none of its submodules.
+    pub fn module(db: &RootDatabase, module: hir::Module) -> SearchScope {
+        let mut entries = IntMap::default();
+
+        let (file_id, range) = {
+            let InFile { file_id, value } = module.definition_source(db);
+            if let Some((file_id, call_source)) = file_id.original_call_node(db) {
+                (file_id, Some(call_source.text_range()))
+            } else {
+                (
+                    file_id.original_file(db),
+                    match value {
+                        ModuleSource::SourceFile(_) => None,
+                        ModuleSource::Module(it) => Some(it.syntax().text_range()),
+                        ModuleSource::BlockExpr(it) => Some(it.syntax().text_range()),
+                    },
+                )
+            }
+        };
+
+        fn split_at_subrange(
+            first: TextRange,
+            second: TextRange,
+        ) -> (TextRange, Option<TextRange>) {
+            let intersect = first.intersect(second);
+            if let Some(intersect) = intersect {
+                let start_range = TextRange::new(first.start(), intersect.start());
+
+                if intersect.end() < first.end() {
+                    (start_range, Some(TextRange::new(intersect.end(), first.end())))
+                } else {
+                    (start_range, None)
+                }
+            } else {
+                (first, None)
+            }
+        }
+
+        if let Some(range) = range {
+            let mut ranges = vec![range];
+
+            for child in module.children(db) {
+                let rng = match child.definition_source(db).value {
+                    ModuleSource::SourceFile(_) => continue,
+                    ModuleSource::Module(it) => it.syntax().text_range(),
+                    ModuleSource::BlockExpr(it) => it.syntax().text_range(),
+                };
+                let mut new_ranges = Vec::new();
+                for old_range in ranges.iter_mut() {
+                    let split = split_at_subrange(old_range.clone(), rng);
+                    *old_range = split.0;
+                    new_ranges.extend(split.1);
+                }
+
+                ranges.append(&mut new_ranges);
+            }
+
+            for range in ranges {
+                entries.insert(file_id, Some(range));
+            }
+        } else {
+            entries.insert(file_id, range);
+        }
```

Comment:
Without this we can have a mismatch between the rustc warning and the assist. Is that ok? I can fix the block expression thing just by ignoring that case in the match.

Reply:
As in an import that is not used in the module but in the child would silence the assist?

Reply:
That is correct. `use_in_submodule_doesnt_count` tests this. If you remove this code then that test will fail.

Reply:
Well, I also just realized that this doesn't work here since SearchScope's entries is a hashmap (so the code here currently only sets the search range to the last one). I think it would be better to just construct the whole module search scope then and have the remove_unused_imports assist split the searchscope accordingly into multiple for each range of the module then (like is done here right now)

Reply:
I moved this to `remove_unused_imports.rs` and made it return a `Vec<SearchScope>`. This function could still easily go in search.rs I think.